pkg/plugins/manager/loader/finder/local.go

```diff
@@ -158,49 +154,79 @@ func (l *Local) readPluginJSON(pluginJSONPath string) (plugins.JSONData, error)
 	return plugin, nil
 }
 
-func (l *Local) getAbsPluginJSONPaths(path string, followDistFolder bool) ([]string, error) {
-	var pluginJSONPaths []string
+func (l *Local) getAbsPluginJSONPaths(rootPath string, class plugins.Class) ([]string, error) {
+	pluginJSONPaths := map[string]struct{}{}
 
 	var err error
-	path, err = filepath.Abs(path)
+	rootPath, err = filepath.Abs(rootPath)
 	if err != nil {
 		return []string{}, err
 	}
 
-	if err = walk(path, true, true, followDistFolder,
+	followDistFolder := true
+	if class == plugins.ClassCore || class == plugins.ClassBundled {
+		followDistFolder = false
+	}
+
+	if err = walk(rootPath, true, true,
 		func(currentPath string, fi os.FileInfo, err error) error {
 			if err != nil {
 				if errors.Is(err, os.ErrNotExist) {
-					l.log.Error("Couldn't scan directory since it doesn't exist", "pluginDir", path, "error", err)
+					l.log.Error("Couldn't scan directory since it doesn't exist", "pluginDir", rootPath, "error", err)
 					return nil
 				}
 				if errors.Is(err, os.ErrPermission) {
-					l.log.Error("Couldn't scan directory due to lack of permissions", "pluginDir", path, "error", err)
+					l.log.Error("Couldn't scan directory due to lack of permissions", "pluginDir", rootPath, "error", err)
 					return nil
 				}
 
 				return fmt.Errorf("filepath.Walk reported an error for %q: %w", currentPath, err)
 			}
 
-			if fi.Name() == "node_modules" {
+			switch {
+			case fi.Name() == "node_modules":
 				return util.ErrWalkSkipDir
-			}
-
-			if fi.IsDir() {
+			case fi.Name() == "dist" && !followDistFolder:
+				return util.ErrWalkSkipDir
+			case fi.IsDir():
 				return nil
-			}
-
-			if fi.Name() != "plugin.json" {
+			case fi.Name() != "plugin.json":
 				return nil
 			}
 
-			pluginJSONPaths = append(pluginJSONPaths, currentPath)
+			pluginJSONPaths[currentPath] = struct{}{}
 			return nil
 		}); err != nil {
 		return []string{}, err
 	}
 
-	return pluginJSONPaths, nil
+	// For external plugins, if we found a plugin directory that contains a "dist" directory, we want to ignore duplicate
+	// plugin.json files found in the parent folder if they exist.
+	// For example, if we found encountered /path/to/plugin/dist/plugin.json, we want to ignore the paths
+	// /path/to/plugin/plugin.json and /path/to/plugin/src/plugin.json in order to prioritize the built
+	// version and avoid loading the plugin twice.
+	if class == plugins.ClassExternal {
+		for pluginJSONPath := range pluginJSONPaths {
+			pluginDir := filepath.Dir(pluginJSONPath)
+
+			// If we found a plugin folder named "dist" at the root level, there should not be any duplicates to remove
+			if filepath.Join(rootPath, "dist") == pluginDir {
+				continue
+			}
```

Comment:
do we need this check? If I am running correctly this in my mind we are preventing from deleting paths like `{rootPath}/plugin.json`  and `{rootPath}/src/plugin.json` which are not something we should worry anyway.

Reply:
Yeah exactly - these files shouldn't exist in this case, but I just wanted to be explicit with this particular case. If it's more confusing I can remove it. But considering we've seen this a few times with badly built/packaged plugins 🤷

Reply:
I don't known, actually if there is a `{rootPath}/plugin.json` or `{rootPath}/src/plugin.json`, it would actually be better to remove them and not load them?

Reply:
😄 I haven't seen those cases TBH. It may be confusing to have this check in that case, so I'll remove this and will actually log when these things are deleted because now it'll be silent

Reply:
At the risk of making things more complicated, I just want to try clarify what I imagine is the right behaviour. Here's examples of what paths will be chosen, depending on what is discovered, per plugin class.

**Core**:
```
✅ /myPlugins/grafana-foo-datasource/plugin.json 
❌ /myPlugins/grafana-foo-datasource/src/plugin.json (should not even exist - especially since these paths are under our control)
❌ /myPlugins/grafana-foo-datasource/dist/plugin.json (can exist due to plugin being built, but should be ignored)
````

```
❌ /myPlugins/grafana-foo-datasource/src/plugin.json (should not exist - especially since these paths are under our control)
❌ /myPlugins/grafana-foo-datasource/dist/plugin.json (can exist due to plugin being built, but should be ignored - Not sure about this one so please confirm @andresmgot )
````

**Bundled**:
```
❌ /myPlugins/grafana-foo-datasource/plugin.json (should not exist - especially since these paths are under our control)
❌ /myPlugins/grafana-foo-datasource/src/plugin.json 
✅ /myPlugins/grafana-foo-datasource/dist/plugin.json (can exist due to plugin being built, and should be prioritized)
````

```
❌ /myPlugins/grafana-foo-datasource/src/plugin.json 
✅ /myPlugins/grafana-foo-datasource/dist/plugin.json (can exist due to plugin being built, and should be prioritized)
````

```
✅  /myPlugins/grafana-foo-datasource/src/plugin.json (should be used if it's the only option)
````

**External**:
```
❌ /myPlugins/grafana-foo-datasource/plugin.json (this is typically what we expect to find)
❌ /myPlugins/grafana-foo-datasource/src/plugin.json (can exist since plugins can theoretically be packaged any way)
✅ /myPlugins/grafana-foo-datasource/dist/plugin.json (this is a legacy decision that we probably need to keep until the next major version)
````

```
✅ /myPlugins/grafana-foo-datasource/plugin.json 
❌ /myPlugins/grafana-foo-datasource/src/plugin.json (can exist since plugins can theoretically be packaged any way)
````

```
❌ /myPlugins/grafana-foo-datasource/src/plugin.json (can exist since plugins can theoretically be packaged any way)
✅ /myPlugins/grafana-foo-datasource/dist/plugin.json (this is a legacy decision that we probably need to keep until the next major version)
````

I wonder can we afford to skip `src/*` 🤔 

In the future all should probably just behave like **Core** and have one way to do things.

Reply:
All the different examples you point out seems correct to me :+1: 

But I see the problem with the complexity here, it would be good to have a simplified policy for all types:

 - if `<pluginID>/dist/plugin.json` exists, it takes precedence
 - if only `<pluginID>/plugin.json` exists, load it
 - ignore any other `<pluginID>/**/plugin.json` (including `src`), that usually implies half-backed plugins that should not be loaded in that state.
 
The only thing that is out of that policy atm are core plugins but there is nothing forcing us to build them in the same directory than its source, so rather than building them in `public/app/plugins/datasource/grafana-testdata-datasource/dist` we could put the output in `plugins-dist/grafana-testdata-datasource` (or similar) in the root folder.

If that makes sense, we can merge this as it is and simplify it later once we adapt core plugins or hold this until we address the other issue. WDYT?

Reply:
> it would be good to have a simplified policy for all types

💯 

>If that makes sense, we can merge this as it is and simplify it later once we adapt core plugins or hold this until we address the other issue. WDYT?

I think what I'll do is just try solve for this case that brought up this whole discussion, by just solving for that - then let's come back and simplify 👍

Reply:
Sorry, I come here to complicate (but later simplify) things :). I forgot that we should actually load the `dist` folder for decoupled plugins, but only if the feature flag is enabled. This allow us to load the external `module.js` rather than the built-in version of the plugin:

https://github.com/grafana/grafana/blob/main/pkg/plugins/manager/loader/assetpath/assetpath.go#L64

So for your original comment:

```
✅ /myPlugins/grafana-foo-datasource/plugin.json 
❌ /myPlugins/grafana-foo-datasource/src/plugin.json (should not even exist - especially since these paths are under our control)
❔ /myPlugins/grafana-foo-datasource/dist/plugin.json (can exist due to plugin being built, but should be ignored if FlagExternalCorePlugins is not enabled, in other case it should load it)
```

This means that once we remove the feature flag, core plugins will behave as external plugins and the `dist` folder takes precedence if it exists.

I need to fix https://github.com/grafana/grafana/pull/77549 since there is where I introduced the bug. Should I wait for any of your PRs to do so or do it before?